### PR TITLE
fix: flaky UnprivilegedEnvironmentSpec

### PR DESCRIPTION
**Pull request by jachro**
/deploy

### CODE REVIEW

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/pages/ProjectPage.scala [hunk 1]
```diff
@@ -361,51 +359,24 @@ class ProjectPage(val projectSlug: String, val namespace: String)
...
+          .find(_.text.trim == "Open")
+      }
 
-      def connectButton(buttonSelector: String)(implicit webDriver: WebDriver): WebElement = eventually {
-        find(
-          cssSelector(buttonSelector)
-        ) getOrElse fail(
-          "First row session Connect button not found"
-        )
+      def maybeGetLogsButton(implicit webDriver: WebDriver): Option[WebElement] = eventually {
```
Comment (travissarles) [inline acceptance-tests/src/test/scala/ch/renku/acceptancetests/pages/ProjectPage.scala:367]: Is this "maybeGetLogsButton" visible to the test writer? I'm not sure if people coming from dynamic languages would necessarily understand that this means Option

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/tooling/Grammar.scala [hunk 2]
```diff
@@ -110,6 +110,15 @@ trait Grammar extends WebElementOps with WebDriverOps with Scripts with Eventual
             s"but gets stuck on $currentUrl for ${(checkInterval * attempt).toSeconds}s"
         }
     }
+
+    @scala.annotation.tailrec
+    def whenUserCanSee(findElement: WebDriver => Option[WebElement], attempt: Int = 1): Unit =
+      if (attempt <= 30 && findElement(webDriver).fold(ifEmpty = false)(_.isDisplayed)) {
+        sleep(1 second)
```
Comment (travissarles) [inline acceptance-tests/src/test/scala/ch/renku/acceptancetests/tooling/Grammar.scala:117]: Should this be more dynamic or is it reliable enough as a hardcoded sleep?

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/workflows/Environments.scala [hunk 3]
```diff
@@ -41,14 +41,49 @@ trait Environments {
...
+
+      And("selects the 'Open in new tab' option")
+      click on projectPage.Sessions.Running.openInNewTabMenuItem
+      sleep(2 seconds)
+    }
+
+    if (webDriver.findTabWithUrlContaining("spawn-pending").nonEmpty) {
+      And("JupyterLab is not up yet")
+      Then("close the window and try again later")
+      // The server isn't up yet. Close the window and try again
```
Comment (travissarles) [inline acceptance-tests/src/test/scala/ch/renku/acceptancetests/workflows/Environments.scala:76]: Is this comment necessary? I understand what's happening from the two lines above

**Review by travissarles — APPROVED** [review]

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/tooling/Grammar.scala (hunk 2 above)
Comment (jachro) [reply acceptance-tests/src/test/scala/ch/renku/acceptancetests/tooling/Grammar.scala:117]: That's always a bit of a problem. It kind of settled down empirically and it's good enough to make the tests pass in all environments. So far I didn't come up with a better dynamic way for this.

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/workflows/Environments.scala (hunk 3 above)
Comment (jachro) [reply acceptance-tests/src/test/scala/ch/renku/acceptancetests/workflows/Environments.scala:76]: I don't know. I just moved this snippet from another place and actually removed quite a few other comments. Maybe this one actually explains something.

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/pages/ProjectPage.scala (hunk 1 above)
Comment (jachro) [reply acceptance-tests/src/test/scala/ch/renku/acceptancetests/pages/ProjectPage.scala:367]: Yes, maybe you are right. I stopped writing non-functional code some time ago and it's somehow natural to me to use the naming and structures. On the other side, I don't know what could be a better and still meaningful name. And maybe it's also a good thing to give our colleagues a bit of a touch of FP ;)

File: acceptance-tests/src/test/scala/ch/renku/acceptancetests/workflows/Environments.scala (hunk 3 above)
Comment (jachro) [reply acceptance-tests/src/test/scala/ch/renku/acceptancetests/workflows/Environments.scala:76]: Ha, ha. Actually, it's interesting why people tend to prefer commenting code over doing more effort while deciding on names.

**Pull request merged**